jobqueue/scheduler/kubernetes.go

```diff
@@ -0,0 +1,683 @@
+		cores = resource.NewMilliQuantity(int64(req.Cores)*100, resource.DecimalSI)
+		ram = resource.NewQuantity(int64(req.RAM)*1024*1024, resource.BinarySI)
+		disk = resource.NewQuantity(int64(req.Disk)*1024*1024, resource.BinarySI)
+	} else {
+		cores = resource.NewMilliQuantity(int64(req.Cores)*1000, resource.DecimalSI)
+		ram = resource.NewQuantity(int64(req.RAM)*1024*1024, resource.BinarySI)
+		disk = resource.NewQuantity(int64(req.Disk)*1024*1024*1024, resource.BinarySI)
+	}
+
+	r := &kubescheduler.Request{
+		RAM:    *ram,
+		Time:   req.Time,
+		Cores:  *cores,
+		Disk:   *disk,
+		Other:  req.Other,
+		CbChan: make(chan kubescheduler.Response),
+	}
+
+	s.Logger.Debug(fmt.Sprintf("Sending request to listener %#v", r))
+	go func() {
+		s.reqChan <- r
+	}()
+
+	s.Logger.Debug("Waiting on reqCheck to return")
+	resp := <-r.CbChan
+
+	s.esmutex.Lock()
+	defer s.esmutex.Unlock()
+
+	if resp.Error != nil {
+		s.Logger.Error(fmt.Sprintf("Requirements check recieved error: %s", resp.Error))
+		s.es = resp.Ephemeral
+		return resp.Error
+	}
+
+	s.Logger.Info("reqCheck returned ok")
+	s.es = resp.Ephemeral
+
+	return resp.Error
+}
+
+// setMessageCallBack sets the given callback function.
+func (s *k8s) setMessageCallBack(cb MessageCallBack) {
+	s.Logger.Info("setMessageCallBack called")
+	s.cbmutex.Lock()
+	defer s.cbmutex.Unlock()
+	s.msgCB = cb
+}
+
+// setBadServerCallBack sets the given callback function.
+func (s *k8s) setBadServerCallBack(cb BadServerCallBack) {
+	s.Logger.Info("setBadServerCallBack called")
+	s.cbmutex.Lock()
+	defer s.cbmutex.Unlock()
+	s.badServerCB = cb
+}
+
+// The controller is passed a callback channel.
+// notifyMessage recieves on the channel
+// if anything is recieved call s.msgCB(msg).
+func (s *k8s) notifyCallBack(callBackChan chan string, badCallBackChan chan *cloud.Server) {
+	s.Logger.Debug("notifyCallBack handler started")
+	for {
+		select {
+		case msg := <-callBackChan:
+			s.Logger.Debug("Callback notification", "msg", msg)
+			if s.msgCB != nil {
+				go s.msgCB(msg)
+			} else {
+				s.logger.Error("No message callback function defined")
+			}
+		case badServer := <-badCallBackChan:
+			s.Logger.Debug("Bad server callback notification", "msg", badServer)
+			if s.badServerCB != nil {
+				go s.badServerCB(badServer)
+			} else {
+				s.logger.Error("No bad server callback function defined")
+			}
+		}
+	}
+
+}
+
+// Delete the namespace when all pods have exited.
+func (s *k8s) cleanup() {
+	s.Logger.Debug("cleanup() Called")
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.cleaned = true
+	err := s.queue.Destroy()
+	if err != nil {
+		s.Warn("cleanup queue destruction failed", "err", err)
+	}
+
+}
+
+// Work out how many pods with given resource requests can be scheduled based on resource requests on the
+// nodes in the cluster.
+func (s *k8s) canCount(req *Requirements) (canCount int) {
+	s.Logger.Debug("canCount Called, returning 100")
+	// 100 is  a big enough block for anyone...
+	return 100
```

Comment:
Calculating when the cluster is full is a difficult task, you've got to calculate the capacity - allocatable from all requests across all pods in each namespace (When I looked, this was how the command `kubectl describe nodes` worked out the number. Hopefully a convenience function may be added later.) I couldn't find how the internal scheduler actually calculates it.

When the cluster is full, runCmd calls just pend. WR does not see them as having failed, the call just waits as it would normally. When there is space somewhere in the cluster, the k8s scheduler will run the jobs. The user sees the job as pending in the web ui and gets a  call back explaining why.